pkg/chain/ethereum/ethutil/adapter_test.go

```diff
@@ -0,0 +1,225 @@
+package ethutil
+
+import (
+	"context"
+	"fmt"
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/keep-network/keep-common/pkg/chain/ethlike"
+	"math/big"
+	"reflect"
+	"testing"
+	"time"
+)
+
+func TestEthlikeAdapter_LatestBlock(t *testing.T) {
+	client := &mockAdaptedEthereumClient{
+		blocks: []*big.Int{
+			big.NewInt(0),
+			big.NewInt(1),
+			big.NewInt(2),
+		},
+	}
+
+	adapter := &ethlikeAdapter{client}
+
+	block, err := adapter.BlockByNumber(context.Background(), nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	expectedBlockNumber := big.NewInt(2)
+	if expectedBlockNumber.Cmp(block.Number) != 0 {
+		t.Errorf(
+			"unexpected last block number\n"+
+				"expected: [%v]\n"+
+				"actual:   [%v]",
+			expectedBlockNumber,
+			block,
+		)
+	}
+}
+
+func TestEthlikeAdapter_SubscribeNewBlocks(t *testing.T) {
+	ctx, cancelCtx := context.WithTimeout(
+		context.Background(),
+		10*time.Millisecond,
+	)
+	defer cancelCtx()
+
+	client := &mockAdaptedEthereumClient{
+		blocks: []*big.Int{
+			big.NewInt(0),
+			big.NewInt(1),
+			big.NewInt(2),
+		},
+	}
+
+	adapter := &ethlikeAdapter{client}
+
+	headerChan := make(chan *ethlike.Header)
+	_, err := adapter.SubscribeNewHead(ctx, headerChan)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	blocks := make([]*big.Int, 0)
+
+loop:
```

Comment:
This test assumes the mocked `SubscribeNewHead` always produces 3 elements in a predefined order. We are testing whether the adapter goroutine works well and passes headers as expected not the `SubscribeNewHead` itself. This is why we stop after getting `3` elements. I see this may not be clear enough so I will try to implement your proposition.

Reply:
See https://github.com/keep-network/keep-common/commit/7c03824d98e1603143f353d23a5c3414199df55d